File: dotnet/src/Grafana.Agento11y/CacheDiagnostics.cs
namespace Grafana.Agento11y;

/// <summary>Reserved generation metadata keys for cache diagnostics (Anthropic cache-diagnosis beta).</summary>
public static class CacheDiagnostics
{
    public const string MissReasonKey = "agento11y.cache_diagnostics.miss_reason";
    public const string MissedInputTokensKey = "agento11y.cache_diagnostics.missed_input_tokens";
    public const string PreviousMessageIdKey = "agento11y.cache_diagnostics.previous_message_id";

    /// <summary>Stamps <c>agento11y.cache_diagnostics.*</c> metadata. Call before <see cref="GenerationRecorder.End"/>.</summary>
    public static void SetCacheDiagnostics(
        GenerationRecorder? recorder,
        string missReason,
        long? missedInputTokens = null,
        string? previousMessageId = null
    ) => recorder?.SetCacheDiagnostics(missReason, missedInputTokens, previousMessageId);
}

File: java/core/src/main/java/com/grafana/agento11y/sdk/CacheDiagnostics.java
package com.grafana.agento11y.sdk;

/** Reserved generation metadata keys for cache diagnostics (Anthropic cache-diagnosis beta). */
public final class CacheDiagnostics {
    public static final String MISS_REASON_KEY = "agento11y.cache_diagnostics.miss_reason";
    public static final String MISSED_INPUT_TOKENS_KEY = "agento11y.cache_diagnostics.missed_input_tokens";
    public static final String PREVIOUS_MESSAGE_ID_KEY = "agento11y.cache_diagnostics.previous_message_id";

    private CacheDiagnostics() {}

    /**
     * Stamps {@code agento11y.cache_diagnostics.*} metadata on a recorder. Call before {@link GenerationRecorder#end()}.
     *
     * @param missedInputTokens pass {@code null} to omit; otherwise written as a decimal string
     * @param previousMessageId pass {@code null} or blank to omit
     */
    public static void setCacheDiagnostics(
            GenerationRecorder recorder,
            String missReason,
            Long missedInputTokens,
            String previousMessageId) {
        if (recorder == null) {
            return;
        }
        recorder.setCacheDiagnostics(missReason, missedInputTokens, previousMessageId);
    }
}
